contracts/apps/accounting/AccountingApp.sol

```diff
@@ -0,0 +1,209 @@
+    }
+
+    function getTransactionState(uint transactionId) constant returns (TransactionState, string) {
+        Transaction memory t = transactions[transactionId];
+        uint tuid = transactionUpdatesRelation[transactionId].length - 1;
+        uint lastTransactionUpdate = transactionUpdatesRelation[transactionId][tuid];
+        TransactionUpdate tu = transactionUpdates[lastTransactionUpdate];
+        return (tu.state, tu.reason);
+    }
+
+    // onlyDAO
+
+    function setDefaultAccountingPeriodSettings(address baseToken, bytes2 ct_hour, bytes2 ct_day, bytes2 ct_month, bytes2 ct_weekday, bytes2 ct_year) onlyDAO {
+        defaultAccountingPeriodSettings.baseToken = baseToken;
+        defaultAccountingPeriodSettings.ct_hour = ct_hour;
+        defaultAccountingPeriodSettings.ct_day = ct_day;
+        defaultAccountingPeriodSettings.ct_month = ct_month;
+        defaultAccountingPeriodSettings.ct_weekday = ct_weekday;
+        defaultAccountingPeriodSettings.ct_year = ct_year;
+    }
+
+    function newIncomingTransaction(address externalAddress, address token, uint256 amount, string reference) onlyDAO {
+        newTransaction(externalAddress, token, amount, reference, TransactionType.Deposit);
+    }
+
+    // Create a new transaction and return the id of the new transaction.
+    // externalAddress is where the transication is coming or going to.
+    function newTransaction(address externalAddress, address token, uint256 amount, string reference, TransactionType _type) onlyDAO {
+        Debug('newTransaction');
+
+        uint tid = transactions.push(Transaction({
+            token: token,
+            amount: amount,
+            // TODO: get base token and exchange rate from oracle
+            baseToken: 0x10000000000001,
+            baseValue: 1,
+            externalAddress: externalAddress,
+            reference: reference,
+            timestamp: now,
+            accountingPeriodId: getCurrentAccountingPeriodId(),
+            _type: _type
+        })) - 1;
+        // All transactions must have at least one state.
+        // To optimize, incoming transactions could go directly to "Suceeded" or "Failed".
+
+        updateTransaction(tid, TransactionState.New, "new");
+    }
+
+    // Create new transactionUpdate for the given transaction id
+    function updateTransaction(uint transactionId, TransactionState state, string reason) onlyDAO returns (uint) {
+        uint tuid = transactionUpdates.push(TransactionUpdate({
+            transactionId: transactionId,
+            state: state,
+            reason: reason,
+            actor: dao_msg.sender
+        })) - 1;
+        transactionUpdatesRelation[transactionId].push(tuid);
+    }
+
+     function approveTransaction(uint transactionId, string reason) onlyDAO {
+        Transaction memory t = transactions[transactionId];
+        if (t._type == TransactionType.Withdrawal){
+            VaultOrgan(dao).transfer(t.token, t.externalAddress, t.amount);
```

Comment:
Excellent point, I will make `updateTransaction` internal and all the `setTransactionX` functions will validate the existing state first. and `approveTransaction` will require the current state to be `pending`. Actually I was thinking of renaming it `approveWithdrawal` because that's a bit more descriptive.